src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                default:
+                    Diagnostics.Assert(false, "ParseText: Unsupported language family.");
+                    break;
+            }
+
+            return null;
+        }
+
+        private SyntaxTree ParseText(SourceText sourceText, ParseOptions parseOptions, string path = "")
+        {
+            switch (Language)
+            {
+                case Language.CSharp:
+                    return CSharpSyntaxTree.ParseText(sourceText, (CSharpParseOptions) parseOptions, path);
+
+                case Language.VisualBasic:
+                    return VisualBasicSyntaxTree.ParseText(sourceText, (VisualBasicParseOptions) parseOptions, path);
+
+                default:
+                    Diagnostics.Assert(false, "ParseText: Unsupported language family.");
+                    break;
+            }
+
+            return null;
+        }
+
+        private CompilationOptions GetDefaultCompilationOptions()
+        {
+            switch (Language)
+            {
+                case Language.CSharp:
+                    var csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
+
+                    return csharpCompilationOptions;
+
+                case Language.VisualBasic:
+                    var vbCompilationOptions = new VisualBasicCompilationOptions(outputKind: OutputAssemblyTypeToOutputKind(OutputType));
+
+                    return vbCompilationOptions;
+
+                default:
+                    Diagnostics.Assert(false, "GetDefaultCompilationOptions: Unsupported language family.");
+                    break;
+            }
+
+            return null;
+        }
+
+        private bool isSourceCodeUpdated(List<SyntaxTree> syntaxTrees, out Assembly assembly)
+        {
+            Diagnostics.Assert(syntaxTrees.Count != 0, "syntaxTrees should contains a source code.");
+
+            _syntaxTreesHash = SyntaxTreeArrayGetHashCode(syntaxTrees);
+
+            if (s_sourceAssemblyCache.TryGetValue(_syntaxTreesHash, out Assembly hashedAssembly))
+            {
+                assembly = hashedAssembly;
+                return false;
+            }
+            else
+            {
+                assembly = null;
+                return true;
+            }
+        }
+
+        private void SourceCodeProcessing()
+        {
+            ParseOptions parseOptions = null;
+            CompilationOptions compilationOptions = null;
+            EmitOptions emitOptions = null;
+
+            if (CompilerOptions != null)
+            {
+                var arguments = DefaultArgsParse(CompilerOptions);
+
+                HandleCompilerErrors(arguments.Errors);
+
+                parseOptions = arguments.ParseOptions;
+                compilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
```

Comment:
What if `-OutputType` is not specified? Shall we respect the `-CompilerOption` or the default value of `-OutputType`?

Reply:
Roslyn [default](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,79) is `OutputKind.ConsoleApplication`. But we should have `Library` default. We don't custom parsing so we have to ignore `CompilerOptions` and always use `OutputType`.

PR description was updated.